src/main/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceStubClassComposer.java

```diff
@@ -245,10 +255,171 @@ protected String getProtoRpcFullMethodName(Service protoService, Method protoMet
     return String.format("google.iam.v1.IAMPolicy/%s", protoMethod.name());
   }
 
-  private LambdaExpr createRequestParamsExtractorClassInstance(Method method) {
-    Preconditions.checkState(
-        method.hasHttpBindings(), String.format("Method %s has no HTTP binding", method.name()));
+  private LambdaExpr createRequestParamsExtractorClassInstance(
+      Method method, List<Statement> classStatements) {
+    List<Statement> bodyStatements = new ArrayList<>();
+    VariableExpr requestVarExpr =
+        VariableExpr.withVariable(
+            Variable.builder().setType(method.inputType()).setName("request").build());
+    TypeNode returnType =
+        TypeNode.withReference(
+            ConcreteReference.builder()
+                .setClazz(Map.class)
+                .setGenerics(TypeNode.STRING.reference(), TypeNode.STRING.reference())
+                .build());
+    Expr returnExpr;
+    // If the google.api.routing annotation is present(even with empty routing parameters),
+    // the implicit routing headers specified in the google.api.http annotation should not be sent
+    if (method.routingHeaders() == null) {
+      returnExpr =
+          addRequestParamsForHttpBindings(method, bodyStatements, requestVarExpr, returnType);
+    } else {
+      returnExpr =
+          addRequestParamsForRoutingHeaders(
+              method, classStatements, bodyStatements, requestVarExpr, returnType);
+    }
 
+    // Overrides extract().
+    // https://github.com/googleapis/gax-java/blob/8d45d186e36ae97b789a6f89d80ae5213a773b65/gax/src/main/java/com/google/api/gax/rpc/RequestParamsExtractor.java#L55
+    return LambdaExpr.builder()
+        .setArguments(requestVarExpr.toBuilder().setIsDecl(true).build())
+        .setBody(bodyStatements)
+        .setReturnExpr(returnExpr)
+        .build();
+  }
+
+  private Expr addRequestParamsForRoutingHeaders(
```

Comment:
Just realized this method returns two values: (1) a return expression for a lambda; and (2) a body for that lambda. I don't like that one is returned as a return value while the other as a mutable argument. Also, `bodyStatements` is in the middle of the argument list, and it's not distinguishable from other read-only arguments.

Reply:
I totally understand where you are coming from and I gave it a lot of thought to make it better. I don't really like the way it is right now as well because the exactly same reason you mentioned, but it's kind of a compromise between reuse as much code as possible and modularize the code. 
That being said, there is still some ways we can improve, we can either pass an ExprBuilder to the method and return void(my preferred approach) or wrap everything we manipulated to a class and return it. I'll update the code and let you know.

Reply:
> ither pass an ExprBuilder to the method and return void(my preferred approach) or wrap everything we manipulated to a class and return it

Yeah, I think either one is better than the current code. Consider renaming the method too.

Reply:
Passing the builder now, also renamed the methods.